ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,155 @@
+              ledger = Big_map.update (from_, token_id) (Some new_bal) store.ledger
+            }
+        | None ->
+            // TODO: How to fail with some value
+            // (failwith ("FA2_INSUFFICIENT_BALANCE", (amt, bal)) : storage)
+            (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+      in updated_store
+  | None ->
+      if (amt = 0n) then store // We allow 0 transfer
+      else (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+
+let credit_to (amt, to_, token_id, store : nat * address * nat * storage): storage =
+  let new_bal =
+    match Big_map.find_opt (to_, token_id) store.ledger with
+      Some bal -> bal + amt
+    | None -> amt
+  in  { store with
+        ledger = Big_map.update (to_, token_id) (Some new_bal) store.ledger
+      }
+
+
+// -----------------------------------------------------------------
+// Transfer
+// -----------------------------------------------------------------
+
+// TODO: how to add to 'where'
+
+let transfer_item (store, ti : storage * transfer_item): storage =
+  let transfer_one (acc, tx : storage * transfer_destination): storage =
+    let valid_from_ = check_sender (ti.from_, acc) in
+    if tx.amount = 0n then
+        acc
+      else (
+        if (tx.token_id = unfrozen_token_id || (tx.token_id = frozen_token_id && sender = store.admin)) then
+          let updated_store =
+            credit_to
+              ( tx.amount, tx.to_, tx.token_id
+              , debit_from (tx.amount, valid_from_, tx.token_id, acc)
+              )
+          in updated_store
+        else if tx.token_id = frozen_token_id then
+          (failwith ("FROZEN_TOKEN_NOT_TRANSFERABLE"): storage)
+        else
+          (failwith ("FA2_TOKEN_UNDEFINED") : storage)
+      )
+  in
+  let result = List.fold transfer_one ti.txs store
+  in result
+
+let transfer (params, store : transfer_params * storage): return =
+  let valid_store = ensure_not_migrated(store) in
+  let updated_store = List.fold transfer_item params valid_store in
+  (([] : operation list), updated_store)
+
+
+// -----------------------------------------------------------------
+// Balance of
+// -----------------------------------------------------------------
+
+[@inline]
+let validate_token_type (token_id : token_id): token_id =
+  if (token_id = unfrozen_token_id || token_id = frozen_token_id) then
+    token_id
+  else (failwith ("FA2_TOKEN_UNDEFINED"): token_id)
+
+let balance_of (params, store : balance_request_params * storage): return =
+  let check_one (res_list, req : balance_response_item list * balance_request_item): balance_response_item list =
+    let valid_token_id = validate_token_type(req.token_id) in
+    match Big_map.find_opt (req.owner, valid_token_id) store.ledger with
+      Some bal ->
+        { request = req; balance = bal} :: res_list
+    | None ->
+        { request = req; balance = 0n} :: res_list
+  in
+  let result = List.fold check_one params.requests ([]: balance_response_item list) in
```

Comment:
No need to rebuild the list manually, we can keep the logic using `List.map` instead of `List.fold`:
```suggestion
  let check_one (req : balance_request_item): balance_response_item =
    let valid_token_id = validate_token_type(req.token_id) in
    match Big_map.find_opt (req.owner, valid_token_id) store.ledger with
      Some bal -> { request = req; balance = bal}
    | None -> { request = req; balance = 0n}
  in
  let result = List.map check_one params.requests in
```
the resulting Michelson code seems to be a little better too.